draft-normington-jsonpath-latest.xml

```diff
@@ -381,57 +381,66 @@ array-slice = [ integer ] ws %x3A ws [ integer ]
-                  </section>       
+                  </section>
                   <section numbered="false" toc="exclude">
                      <name>Semantics</name>
-                     <t>An array index selects any element of an array which is indexed by the integer index.
-                        It selects no nodes from a node which is not an array.
-                     </t>
-                     <t>A non-negative index <tt>i</tt> selects an element of an array of length <tt>len</tt> if and only if <tt>0 &lt;= i &lt; len</tt>.
-                        The index <tt>0</tt> selects the first element of a non-empty array and, providing <tt>i &lt; len - 1</tt>, index <tt>i + 1</tt>
-                        selects the element immedately after (in array order) the element selected by index <tt>i</tt>.
-                     </t>
-                     <t>A negative index <tt>j</tt> selects an element of an array of length <tt>len</tt> if and only if <tt>0 &lt;= j + len &lt; len</tt>,
-                        in which case it selects the same element as the non-negative index <tt>j + len</tt>.
-                     </t>
-                     <t>An array slice selects any elements of an array which are indexed by the members of the slice,
-                        as defined below, in the same order as the slice.
-                        It selects no nodes from a node which is not an array.
-                     </t>
-                     <t>The members of a slice depend on whether the step value is positive, negative, or zero as follows. If the step value is
-                        omitted, it defaults to <tt>1</tt>.
-                     </t>
-                     <t>For positive integer <tt>step</tt> and integers <tt>start</tt> and <tt>end</tt>, the array slice
-                        <tt>start:end:step</tt> consists of the indices starting at <tt>start</tt> and increasing repeatedly
-                        by <tt>step</tt> while the resultant index is less than <tt>end</tt>.
-                        The indices of the slice are the same as the values of i iterated over by this <tt>for</tt> loop in C:
-                     </t>
+                     <t>A slice expression produces a new array containing a subset of the elements of the input array.</t>
+                     <t>A slice is defined by the two bounds, <tt>start</tt> and <tt>end</tt>,
+                        and a <tt>step</tt> which governs the direction of iteration and the iteration stride.
+                        Each of those parameters is optional.</t>
+                     <t>Negative bounds are interpreted as offsets from the end of the array, namely <tt>b + len</tt>.
+                        This remapping is performed before any other evaluation step.</t>
+                     <t>A slice bound <tt>b</tt> is considered valid if <tt>0 &lt;= b &lt; len</tt>.
+                        An invalid bound is treated as an omitted bound. Omitted bounds take a default value
```

Comment:
I'm not keen on the "invalid" terminology, as this connotes an error or mistake. Such bounds are perfectly valid. We can either describe how they are clamped, depending on the step value. Or we could allow them to default to +/- infinity and make it clear how very large positive and negative explicit (including defaulted) bounds are interpreted.